src/lava/utils/profiler.py

```diff
@@ -12,3 +12,122 @@
 elementary operation has a defined execution time and energy cost, which is
 used in a performance model to calculate execution time and energy.
 """
+import typing as ty
+from abc import ABC
+import types
+import numpy as np
+from lava.magma.core.process.process import AbstractProcess
+from lava.magma.core.run_conditions import AbstractRunCondition
+from lava.magma.core.run_configs import RunConfig
+from lava.magma.runtime.runtime import Runtime
+from lava.magma.core.process.message_interface_enum import ActorType
+from lava.magma.compiler.compiler import Compiler
+from lava.magma.core.resources import Loihi1NeuroCore, Loihi2NeuroCore
+
+
+class LavaProfiler(ABC):
+    """TBD"""
+
+    def __init__(self, start: int = 0, end: int = 0,
+                 bin_size: int = 1, buffer_size: int = 1000):
+        self.start = start
+        self.end = end
+        self.bin_size = bin_size
+        self.buffer_size = buffer_size
+        self.loihi_flag = False  # Exchange with list of resources
+
+    def profile(self, proc: AbstractProcess):
```

Comment:
I still think that the user API would be better if the user could simply define a profiler without having to change his line
proc.run(...),
i.e. the interface that PP coded here. But I also see Andreas point: Overwriting the run function feels bad. In particular, what happens if we ever change the general run method of Processes? Then we may forget to change it in the Profiler, and it will stop working.
What does the Profiler actually do under the hood? Does it just count how often the internal run_spk method is called, and how often spikes are being sent? In that case, could we 
1) ask people to just add a decorator @count to count runs in front of the run_spk function:
class MyProcModel(list):
    @counter
    def run_spk(self, *args, **kwargs):
        ...
with 
def counter(f):
    def count(*args, **kwargs):
        count.calls += 1
        return f(*args, **kwargs)
    count.calls = 0
    return count
This decorator would mean that we don't need to rewrite the whole run method.
2) subscribe to the port of the Process to read its sent spikes?

To me, that would be way less invasive. But I know too little about what's going on under the hood to say if anything like that would work.

Reply:
Both versions are slide by slide in the PowerPoint ;)

@phstratmann We need to do quite a few things under the hood, especially modifying the compilation process (the sketch of what we need to do is in form of comments and mock methods in this PR).
We do not really count how often run_spk is called.